Objects/unicodeobject.c

```diff
@@ -15063,9 +15175,74 @@ _PyUnicode_InitTypes(PyInterpreterState *interp)
     return _PyStatus_ERR("Can't initialize unicode types");
 }
 
-
 void
-_PyUnicode_InternInPlace(PyInterpreterState *interp, PyObject **p)
+_PyUnicode_InternStatic(PyInterpreterState *interp, PyObject **p)
+{
+    PyObject *s = *p;
+    assert(s != NULL);
+    assert(_PyUnicode_CHECK(s));
+    assert(_PyUnicode_STATE(s).statically_allocated);
+    assert(_Py_IsImmortal(s));
+
+    switch (PyUnicode_CHECK_INTERNED(s)) {
+        case SSTATE_NOT_INTERNED:
+            break;
+        case SSTATE_INTERNED_IMMORTAL_STATIC:
+            return;
+        default:
+            Py_FatalError("_PyUnicode_InternStatic called on wrong string");
+    }
+
+#ifdef Py_DEBUG
+    /* We must not add process-global interned string if there's already a
+     * per-interpreter interned_dict, which might contain duplicates.
+     * Except "short string" singletons: those are special-cased. */
+    PyObject *interned = get_interned_dict(interp);
+    assert(interned == NULL || unicode_is_singleton(s));
+#ifdef Py_GIL_DISABLED
+    // In the free-threaded build, don't allow even the short strings.
+    assert(interned == NULL);
+#endif
+#endif
+
+    /* Look in the global cache first. */
+    PyObject *r = (PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s);
+    /* We should only init each string once */
+    assert(r == NULL);
+    /* but just in case (for the non-debug build), handle this */
+    if (r != NULL && r != s) {
+        assert(_PyUnicode_STATE(r).interned == SSTATE_INTERNED_IMMORTAL_STATIC);
+        Py_SETREF(*p, Py_NewRef(r));
+        assert(_PyUnicode_CHECK(r));
+        return;
+    }
```

Comment:
This function steals a reference, right?  If so, I'm pretty sure we need to decref `s`:

```suggestion
    if (r != NULL && r != s) {
        assert(_PyUnicode_STATE(r).interned == SSTATE_INTERNED_IMMORTAL_STATIC);
        Py_SETREF(*p, Py_NewRef(r));
        assert(_PyUnicode_CHECK(r));
        Py_DECREF(s);
        return;
    }
```

Reply:
We've asserted that `r` and `s` are immortal. And this is code for when an invariant is broken; the ref total doesn't really matter.
But, adding `Py_DECREF` is better than adding a comment to explain that :)